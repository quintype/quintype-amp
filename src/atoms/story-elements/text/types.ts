export interface TextProps {
  externalLink?: boolean;
}
