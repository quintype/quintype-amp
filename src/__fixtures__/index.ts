import { config, publisherConfig, ampConfig, configOpts } from "./config.fixture";
import { textStory } from "./text-story.fixture";
import { storyWithManyJsEmbeds } from "./story-with-jsembeds.fixture";
import { relatedStories } from "./related-stories";
import { seo } from "./seo.fixture";
import { allElementsStory } from "./all-element-story.fixture";
import { liveBlogStory } from "./live-blog-story.fixture";

export {
  allElementsStory,
  config,
  publisherConfig,
  ampConfig,
  textStory,
  storyWithManyJsEmbeds,
  relatedStories,
  configOpts,
  seo,
  liveBlogStory
};
