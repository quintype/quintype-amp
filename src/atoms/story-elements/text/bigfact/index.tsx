import { Bigfact } from "./bigfact";
export { Bigfact };
