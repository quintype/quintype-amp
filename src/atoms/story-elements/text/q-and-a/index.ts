import { QAndA } from "./q-and-a";
export { QAndA };
