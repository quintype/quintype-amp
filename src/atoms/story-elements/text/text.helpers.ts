import get from "lodash.get";
import { parse } from "node-html-parser";

export const conditionExternalLinks = ({ text, config }) => {
  // finds external links and adds rel=”nofollow” target="_blank" to them

  const sketchesHost = get(config, ["publisherConfig", "sketches-host"]);
  const escapedSketchesHost = escapeRegex(sketchesHost);
  const regex = new RegExp(`^((?!${escapedSketchesHost}).)*$`);
  const domTree = parse(text);
  const anchorsArr = domTree.querySelectorAll("a");
  let accumulator = text;
  anchorsArr.forEach((el) => {
    const href = el.rawAttributes.href;
    if (regex.test(href)) {
      const escapedHref = escapeRegex(href);
      accumulator = accumulator.replace(
        new RegExp(`href="${escapedHref}"`),
        `href="${href}" rel=”nofollow” target="_blank"`
      );
    }
  });
  return accumulator;
};

function escapeRegex(str: string) {
  return str.replace(/[-\/\\^$*+?.()|[\]{}]/g, "\\$&");
}
