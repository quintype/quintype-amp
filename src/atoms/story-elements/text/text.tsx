import React from "react";
import { StoryElement } from "../../../types/story";
import styled from "styled-components";

export interface TextProps {
  element: StoryElement;
  externalLink?: boolean;
}

const BaseText = ({ element, externalLink }: TextProps) => {
  let text = element.text || "";
  if (externalLink) {
    text = text.replace(/<a/g, '<a target="_blank"');
  }
  return <div dangerouslySetInnerHTML={{ __html: text }} />;
};

const Text = styled(BaseText)`
  color: ${(props) => props.theme.color.mono6};
  font-size: ${(props) => props.theme.font.size.m};
  font-family: ${(props) => props.theme.font.family.primary};
  line-height: ${(props) => props.theme.font.lineHeight.level3};
  p {
    margin: 0 0 ${(props) => props.theme.spacing.xs} 0;
  }
  a {
    color: ${(props) => props.theme.color.primaryColor};
  }
  ul {
    list-style: disc;
    margin: 0;
    padding: 0 0 0 ${(props) => props.theme.spacing.l};
  }

  ol {
    list-style: decimal;
    margin: 0;
    padding: 0 0 0 ${(props) => props.theme.spacing.l};
  }

  h2 {
    font-size: ${(props) => props.theme.font.size.huge};
    line-height: ${(props) => props.theme.font.lineHeight.level5};
    font-family: ${(props) => props.theme.font.family.primary};
    margin: 0 0 ${(props) => props.theme.spacing.s} 0;
  }

  h3 {
    font-size: ${(props) => props.theme.font.size.xl};
    line-height: ${(props) => props.theme.font.lineHeight.level5};
    font-family: ${(props) => props.theme.font.family.primary};
    margin: 0 0 ${(props) => props.theme.spacing.s} 0;
  }
`;

export { Text, BaseText };
