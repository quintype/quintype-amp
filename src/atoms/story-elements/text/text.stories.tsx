import * as React from "react";
import { storiesOf } from "@storybook/react";
import { Text } from "./text";
import { config, textStory } from "../../../__fixtures__";
import { Layout } from "../../layout";

const sampleTextElement = {
  id: "1",
  type: "text",
  subtype: null,
  text:
    "<p> It's a baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed. <a href='www.google.com'>Search online</a> </p><p>Tumeric jianbing godard shaman lomo blog blue bottle cloud bread vaporware whatever vape drinking vinegar austinhashtag adaptogen. Post-ironic offal irony leggings brunch. Flexitarian hexagon banh mi, yuccie unicorn offal smallbatch af vice poke gluten-free brooklyn.</p><h2>Un ordered list</h2><p> baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed.</p><ul><li>One</li><li>Two</li><li>Three</li></ul><h3>Ordered List</h3><p> baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed.</p><ol><li>One</li><li>Two</li><li>Three</li></ol>"
};
const sampleSummaryElement = {
  id: "2",
  type: "text",
  subtype: "summary",
  text:
    "<p> It's a baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed. </p>"
};

storiesOf("Text", module)
  .addDecorator((story) => (
    <Layout story={textStory} config={config}>
      {story()}
    </Layout>
  ))
  .add("Text", () => <Text element={sampleTextElement} />)
  .add("Text with external link", () => <Text element={sampleTextElement} externalLink={true} />)
  .add("Summary element", () => <Text element={sampleSummaryElement} />);
