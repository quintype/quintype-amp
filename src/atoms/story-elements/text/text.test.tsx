import React from "react";
import { Text, TextBase } from "./text";
import { shallow } from "enzyme";
import { config, textStory } from "../../../__fixtures__";

const sampleTextElement = {
  id: "1",
  type: "text",
  subtype: null,
  text:
    "<p> It's a baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed. <a href='https://www.google.com/'>Search online</a> </p><p>Tumeric jianbing godard shaman lomo blog blue bottle cloud bread vaporware whatever vape drinking vinegar austinhashtag adaptogen. Post-ironic offal irony leggings brunch. Flexitarian hexagon banh mi, yuccie unicorn offal smallbatch af vice poke gluten-free brooklyn.</p><h2>Un ordered list</h2><p> baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed.</p><ul><li>One</li><li>Two</li><li>Three</li></ul><h3>Ordered List</h3><p> baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed.</p><ol><li>One</li><li>Two</li><li>Three</li></ol>"
};

describe("Text", () => {
  it("should render text", () => {
    const wrapper = shallow(<Text element={sampleTextElement} />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render text and call storyElementRender prop when passed to opts", () => {
    const textElementRender = jest.fn();
    const modifiedConfig = { ...config, opts: { ...config.opts, storyElementRender: { textElementRender } } };
    const wrapper = shallow(<TextBase element={sampleTextElement} story={textStory} config={modifiedConfig} />);
    expect(textElementRender.mock.calls.length).toBe(1);
    expect(wrapper.find("p").length).toBe(0);
  });
});
