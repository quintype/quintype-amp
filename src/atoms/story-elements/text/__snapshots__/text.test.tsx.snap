// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render text 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p> It's a baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed. <a href='www.google.com'>Search online</a> </p><p>Tumeric jianbing godard shaman lomo blog blue bottle cloud bread vaporware whatever vape drinking vinegar austinhashtag adaptogen. Post-ironic offal irony leggings brunch. Flexitarian hexagon banh mi, yuccie unicorn offal smallbatch af vice poke gluten-free brooklyn.</p><h2>Un ordered list</h2><p> baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed.</p><ul><li>One</li><li>Two</li><li>Three</li></ul><h3>Ordered List</h3><p> baby put a bird on it cliche DIY swag photo booth portland helvetica slow-carb sartorial affogato drinking vinegarwhatever. Microdosing dreamcatcher coloring book tousled tattooed.</p><ol><li>One</li><li>Two</li><li>Three</li></ol>",
    }
  }
/>
`;
